.adr/0001-offchain-protocols-as-flowcharts.md

```diff
@@ -0,0 +1,42 @@
+## 0001 -- Off-chain protocols as Flowcharts
+
+## Status
+
+Accepted
+
+## Context
+
+We require a consistent approach to encoding off-chain state channel protocols which:
+
+- tames complexity
+- reduces data duplication and synchronization issues
+- allows for "restartability" of the protocol if it is interrupted before completing
+- allows for easy testing
+
+## Decision
+
+An objective encodes a state chart (or state machine), but in a slightly non-standard manner:
+
+- The **enumerable state** of the objective consists of several "pause points" where the current peer cannot progress without external input `X`. Hence we call theses states "Waiting For `X`".
+- The **enumerable state** of the objective is not stored anywhere: it is computed from the **extended state**.
+- The **extended state** of the objective (sometimes known as the "context") contains potentially infinite data such as off-chain messages, signatures and so on.
+- State transitions are triggered by **events** in one of two ways:
+  - new information comes to light and the extended state is **updated**
+  - the objective is **cranked** -- driven towards completion -- and side effects are generated and returned. This may also update the extended state.
+- The objective progress can be thought of as a flow chart (rather than a conventional state chart). When a blocking condition is identified, the state transition method call returns "early". As more progress is made, the execution reaches deeper into the method until eventually it reaches the end and the objective is complete.
```

Comment:
It might be worth drawing a parallel to eg. asynchronous javascript code. Progress continues until the engine needs to await a promise, and is restarted again once the promise is fulfilled or rejected.